front/app/src/js/components/game/Engine/_GameSocketIO.js

```diff
@@ -86,167 +66,226 @@ export class _GameSocketIO {
-      }
-      console.log('prepare_ball_for_match received');
+      await this.#prepareBallForMatchEventHandler(data);
+    });
 
-      if (!this.#gameHasStarted) {
-        this.#gameHasStarted = true;
-        this.#engine.startListeningForKeyHooks();
-        console.log('Game is starting'); // TODO remove me
-        // TODO remove waiting for other players message
-      }
+    this.#socketIO.on('update_ball', async (data) => {
+      await this.#updateBallEventHandler(data);
+    });
 
-      const match = this.#engine.scene
-          .getMatchFromLocation(data['match_location']);
-      match.prepare_ball_for_match(
-          ServerTimeFixer.fixServerTime(data['ball_start_time']),
-          data['ball_movement'],
-      );
+    this.#socketIO.on('player_won_match', async (data) => {
+      await this.#playerWonMatchEventHandler(data);
     });
 
-    this.#socketIO.on('update_ball', async (data) => {
-      while (!(this.#engine.scene instanceof Scene)) {
-        await sleep(50);
-      }
-      console.log('update_ball received');
-
-      const match = this.#engine.scene
-          .getMatchFromLocation(data['match_location']);
-      const movement = data['movement'];
-      match.setBallMovement(movement);
-      match.setBallPosition(data['position']);
-
-      const paddle = movement['x'] < 0. ?
-          match.players[0].paddle :
-          match.players[1].paddle;
-      const timeAtUpdate = ServerTimeFixer
-          .fixServerTime(data['time_at_update']);
-      const now = Date.now();
-      if (timeAtUpdate >= now) {
-        return;
-      }
-      const timeDelta = (now - timeAtUpdate) / 1000.;
+    this.#socketIO.on('player_scored_a_point', async (data) => {
+      await this.#playerScoredAPointEventHandler(data);
+    });
+
+    this.#socketIO.on('game_over', async (data) => {
+      await this.#gameOverEventHandler(data);
+    });
+
+    this.#socketIO.connect();
+  }
+
+  requestTimeSync(currentTime) {
+    this.emit('request_time_sync', currentTime);
+  }
+
+  async #connectErrorHandler(jsonString, URI) {
+    let error;
+    try {
+      error = JSON.parse(jsonString.message);
+    } catch {
+      ToastNotifications.addErrorNotification('Server connection error');
+      this.disconnect();
+      this.#reconnectAttempts++;
+      await sleep(2000);
+      await this.init(URI);
+      return;
+    }
+    if (error['status_code'] === 0 || error['status_code'] === 1) {
+      ToastNotifications.addErrorNotification(error['message']);
+      getRouter().redirect('/');
+      return;
+    } else {
+      console.error('Connection error:', error['message']);
+      ToastNotifications.addErrorNotification(`Connection error: ${error['message']}`);
+      this.disconnect();
+      this.#reconnectAttempts++;
+      await sleep(2000);
+      await this.init(URI);
+    }
+  }
+
+  async #sceneEventHandler(data) {
+    while (!ServerTime.isTimeSynced) {
+      await sleep(5);
+    }
+    const scene = new Scene();
```

Comment:
Wait for the first time sync before creating the scene as some time related data will need adjusting